src/linker/Linker/LinkerILProcessor.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+namespace Mono.Linker
+{
+	class LinkerILProcessor
+	{
+		readonly ILProcessor _ilProcessor;
+
+		Collections.Generic.Collection<Instruction> Instructions => _ilProcessor.Body.Instructions;
+
+		internal LinkerILProcessor (MethodBody body)
+		{
+			_ilProcessor = body.GetILProcessor ();
+		}
+
+		public void Emit (OpCode opcode) => Append (_ilProcessor.Create (opcode));
+		public void Emit (OpCode opcode, TypeReference type) => Append (_ilProcessor.Create (opcode, type));
+		public void Emit (OpCode opcode, MethodReference method) => Append (_ilProcessor.Create (opcode, method));
+		public void Emit (OpCode opcode, VariableDefinition variable) => Append (_ilProcessor.Create (opcode, variable));
+
+		public void Emit (OpCode opcode, string value) => Append (_ilProcessor.Create (opcode, value));
+		public void InsertBefore (Instruction target, Instruction instruction)
+		{
+			// So when inserting before, all pointers have to be updated to the new "before" instruction.
+			RedirectScopeStart (target, instruction);
+			_ilProcessor.InsertBefore (target, instruction);
+		}
+
+		// When inserting after, no redirection is necessary since it will naturally be "appended" to the same blocks
+		// as the target instruction.
+		public void InsertAfter (Instruction target, Instruction instruction)
+		{
+			RedirectScopeEnd (target, instruction);
+			_ilProcessor.InsertAfter (target, instruction);
+		}
+
+		public void Append (Instruction instruction)
+		{
+			Instruction lastInstruction = Instructions.Count == 0 ? null : Instructions[Instructions.Count - 1];
+			RedirectScopeEnd (lastInstruction, instruction);
+			_ilProcessor.Append (instruction);
+		}
+
+		public void Replace (Instruction target, Instruction instruction)
+		{
+			RedirectScopeStart (target, instruction);
+			RedirectScopeEnd (target, instruction);
+			ReplaceInstruction (target, instruction);
+			_ilProcessor.Replace (target, instruction);
+		}
+
+		public void Replace (int index, Instruction instruction) => Replace (_ilProcessor.Body.Instructions[index], instruction);
+
+		public void Remove (Instruction instruction)
+		{
+			int index = _ilProcessor.Body.Instructions.IndexOf (instruction);
+			if (index == -1)
+				throw new ArgumentOutOfRangeException (nameof (instruction));
+
+			Instruction nextInstruction = Instructions.Count == index + 1 ? null : Instructions[index + 1];
+			Instruction previousInstruction = index == 0 ? (Instructions.Count == 1 ? null : Instructions[index + 1]) : Instructions[index - 1];
```

Comment:
Is the index+1 case for previousInstruction intentional?

Reply:
Yes - if we're removing the first instruction, we have to re-point to "something", in this case the previous and next will be the same, but there's no other way really.

Reply:
I'm a little confused by this - if removing the last instruction, we are redirecting scope starts to null. Shouldn't removing the first instruction be symmetric, and redirect scope ends to null?

Reply:
You're right. I think the correct behavior here is actually to remove the exception handler if this happens - basically if we try to redirect the "Start" of the exception handler to null, we should remove it.
That said this will never happen in the linker I think, branch removal handles removing of exception handlers on its own. So I would rather leave this as is - and just point it to null. Again - purposefully getting into invalid state, but so be it.

Reply:
Agreed - removing the handler might be reasonable, but I think it's ok not to do too much "magic" here if you are getting into such weird states. Same for the case where we remove the last instruction and it's a branch target - in this case you just need to know what you're doing.